src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs

```diff
@@ -262,60 +212,11 @@ private static void ExpirationTokensExpired(object obj)
             }, obj, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
         }
 
-        private void DetachTokens()
-        {
-            // _expirationTokenRegistrations is not checked for null, because AttachTokens might initialize it under lock
-            lock (_lock)
-            {
-                IList<IDisposable> registrations = _expirationTokenRegistrations;
-                if (registrations != null)
-                {
-                    _expirationTokenRegistrations = null;
-                    for (int i = 0; i < registrations.Count; i++)
-                    {
-                        IDisposable registration = registrations[i];
-                        registration.Dispose();
-                    }
-                }
-            }
-        }
-
-        internal void InvokeEvictionCallbacks()
-        {
-            if (_postEvictionCallbacks != null)
-            {
-                Task.Factory.StartNew(state => InvokeCallbacks((CacheEntry)state), this,
-                    CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
-            }
-        }
-
-        private static void InvokeCallbacks(CacheEntry entry)
-        {
-            IList<PostEvictionCallbackRegistration> callbackRegistrations = Interlocked.Exchange(ref entry._postEvictionCallbacks, null);
-
-            if (callbackRegistrations == null)
-            {
-                return;
-            }
-
-            for (int i = 0; i < callbackRegistrations.Count; i++)
-            {
-                PostEvictionCallbackRegistration registration = callbackRegistrations[i];
-
-                try
-                {
-                    registration.EvictionCallback?.Invoke(entry.Key, entry.Value, entry.EvictionReason, registration.State);
-                }
-                catch (Exception e)
-                {
-                    // This will be invoked on a background thread, don't let it throw.
-                    entry._cache._logger.LogError(e, "EvictionCallback invoked failed");
-                }
-            }
-        }
+        internal void InvokeEvictionCallbacks() => _tokens?.InvokeEvictionCallbacks(this);
 
         // this simple check very often allows us to avoid expensive call to PropagateOptions(CacheEntryHelper.Current)
-        internal bool CanPropagateOptions() => _expirationTokens != null || AbsoluteExpiration.HasValue;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // added based on profiling
+        internal bool CanPropagateOptions() => (_tokens != null && _tokens.CanCopyTokens()) || AbsoluteExpiration.HasValue;
```

Comment:
(nit) Can we rename `_tokens.CanCopyTokens()` => `_tokens.CanPropagateTokens()`? It feels odd calling these operations 2 different words (`Propagate` and `Copy`) in the same line.